### PR TITLE
unit tests for functions in helper_linux.go file

### PR DESCRIPTION
**Pull request by vjayaramrh**
Signed-off-by: Vishwanath Jayaraman <vjayaram@redhat.com>

**- What this PR does and why is it needed**

This PR adds unit tests for the functions in helper_linux.go and makes refactor where necessary

**- Special notes for reviewers**

This PR Defines two interfaces for testability in the pkg/cni/helper_linux.go file
- CNIPluginLibOps interface aggregates the calls to the functions that are defined in the library github.com/containernetworking/plugins/pkg/ip
- SriovNetLibOps interface aggregates the calls to the functions that are defined in the library "github.com/Mellanox/sriovnet"
The functions within the helper_linux.go file are refactored to use the methods in the above referenced interface

A mocks directory created in pkg/cni/mocks  contains the below mock (generated using mockery —> https://github.com/vektra/mockery) files that are used in unit tests
- CNIPluginLibOps.go
- SriovNetLibOps.go

Mock (generated using vektra/mockery) for the interface `Result` defined in file `github.com/containernetworking/cni/pkg/types\types.go` is located at `pkg/testing/mocks/containernetworking/cni/pkg/types/Result.go`. This mock is used in unit tests.

Mock (generated using vektra/mockery) for the interface `NetNS` defined in file `github.com/containernetworking/plugins/pkg/ns/ns_linux.go’ is located at `pkg/testing/mocks/containernetworking/plugins/pkg/ns/NetNS.go`. This mock is used in unit tests

The interface definition `NetLinkOps` in `go-controller/pkg/util/net_linux.go` file was updated to add below methods that are used in the helper_linux.go 
-LinkSetDown(link netlink.Link) error
-LinkSetName(link netlink.Link, newName string) error
-LinkSetUp(link netlink.Link) error
-LinkSetNsFd(link netlink.Link, fd int) error
-LinkSetHardwareAddr(link netlink.Link, hwaddr net.HardwareAddr) error
-LinkSetMTU(link netlink.Link, mtu int) error
-LinkSetTxQLen(link netlink.Link, qlen int) error

NOTE: Unable to author Unit tests for `func (pr *PodRequest) ConfigureInterface(namespace string, podName string, ifInfo *PodInterfaceInfo)` without refactoring the `type PodRequest struct` having a `type NetNS` field, will look into this at a later time.

**- How to verify it**

Click on the coverage/coveralls link as part of CI run and click on file named helper_linux.go under CHANGED tab 

**- Description for the changelog**

### CODE REVIEW

File: go-controller/pkg/cni/helper_linux_test.go [hunk 1]
```diff
@@ -0,0 +1,987 @@
...
+				}
+				for _, ret := range item.RetArgList {
+					call.ReturnArguments = append(call.ReturnArguments, ret)
+				}
+				call.Once()
+			}
+			err := moveIfToNetns(tc.inpIfaceName, tc.inpNetNs)
+			t.Log(err)
+			if tc.errMatch != nil {
+				assert.Contains(t, err.Error(), tc.errMatch.Error())
```
Comment (girishmg) [inline go-controller/pkg/cni/helper_linux_test.go:192]: Can we just it errors out or not? The error string is not a contract between the API caller and the API implementor. Also, you are only checking some part of the error message. So, it is better to just check if the API failed or not.

File: go-controller/pkg/cni/helper_linux_test.go [hunk 2]
```diff
@@ -0,0 +1,987 @@
...
+			} else {
+				assert.Nil(t, err)
+			}
+			mockNetLinkOps.AssertExpectations(t)
```
Comment (girishmg) [inline go-controller/pkg/cni/helper_linux_test.go:196]: do you also need

`mockNetNS.AssertExpectations(t)`?

File: go-controller/pkg/cni/helper_linux_test.go [hunk 3]
```diff
@@ -0,0 +1,987 @@
...
+				call.Once()
+			}
+			err := setupNetwork(tc.inpLink, tc.inpPodIfaceInfo)
+			t.Log(err)
+			if tc.errMatch != nil {
+				assert.Contains(t, err.Error(), tc.errMatch.Error())
+			} else {
+				assert.Nil(t, err)
+			}
+			mockNetLinkOps.AssertExpectations(t)
```
Comment (girishmg) [inline go-controller/pkg/cni/helper_linux_test.go:363]: L359: Same question as before. I think it is sufficient to check if `err != nil` and not the contents of the error itself.

File: go-controller/pkg/cni/helper_linux_test.go [hunk 4]
```diff
@@ -0,0 +1,987 @@
...
+	cni_ns_mocks "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/testing/mocks/containernetworking/plugins/pkg/ns"
+	netlink_mocks "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/testing/mocks/vishvananda/netlink"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	util_mocks "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util/mocks"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/vishvananda/netlink"
+	"net"
+	"testing"
+)
```
[the remark this answers is not in the document]
Comment (vjayaramrh) [reply go-controller/pkg/cni/helper_linux_test.go:23]: sure, will fix it

File: go-controller/pkg/cni/helper_linux_test.go (hunk 1 above)
Comment (vjayaramrh) [reply go-controller/pkg/cni/helper_linux_test.go:192]: My initial approach was like what you were suggesting, however, the current approach taken is after I got a comment (https://github.com/ovn-org/ovn-kubernetes/pull/1474#discussion_r447723068) from @danwinship to one of my PRs related to tests for annotations files in pkg/util. If we just checked for error only, we could still accidentally make tests pass and not catch any changes made to the code.

**Comment by vjayaramrh** [comment]
> @vjayaramrh may be you have answered this before, so sorry if it is a repeat question.
> 
> Why are certain mocks under the main `go-controller/pkg/testing` folder whilst some mocks are under the source code folder itself.
@girishmg Yes, I had answered a similar question in one of my PR's earlier at https://github.com/ovn-org/ovn-kubernetes/pull/1371#discussion_r434186468

In a nutshell, 
The mocks that are generated from the interfaces in the vendor directory are in `go-controller/pkg/testing` folder while the  the interfaces that are defined in our project pkg (ie. the go-controller/pkg/*) have their mocks in their respective directories.

File: go-controller/pkg/cni/helper_linux_test.go (hunk 2 above)
Comment (vjayaramrh) [reply go-controller/pkg/cni/helper_linux_test.go:196]: good catch, thanks, will fix it

File: go-controller/pkg/cni/helper_linux_test.go (hunk 3 above)
Comment (vjayaramrh) [reply go-controller/pkg/cni/helper_linux_test.go:363]: same response as above

**Pull request merged**